### ISSUE TITLE
Is there a way to implement a function that takes callback?

### ISSUE THREAD

**Issue by newhoggy** [polysemy-research/polysemy#216]
I'd like to implement a function that looks like this:

```
awsMkEnv'
    :: Member (Embed IO) r
    => Region
    -> (LogLevel -> LBS.ByteString -> Sem r ())
    -> Sem r a
awsMkEnv' region cb = ???
```

It should call `mkEnv :: Region -> (LogLevel -> ByteString -> IO ()) -> IO Env`.

Is that possible?

Thanks!

**Comment by KingoftheHomeless** [comment]
This will become possible once `Final` from `polysemy-zoo` is moved to `polysemy` proper!

```haskell
awsMkEnv'
    :: Member (Final IO) r
    => Region
    -> (LogLevel -> LBS.ByteString -> Sem r ())
    -> Sem r Env
awsMkEnv' region cb = withStrategicToFinal $ do
  cb' <- bindS (uncurry cb)
  s    <- getInitialStateS
  liftS $ mkEnv region $ \loglvl str -> void $ cb' ((loglvl, str) <$ s)
```

At the moment, I recommend you instead encode that action as an effect and then use an interpreter which takes an explicit lowering function in order to interpret that effect.

**Comment by KingoftheHomeless** [comment]
Wait, that's exactly what you did with #214 ! So is your question if there's a way to not need the lowering function? Because if so, `Final` is exactly what you're looking for!

**Comment by KingoftheHomeless** [comment]
Writing an `Aws` interpreter using `Final` would look very similar:

```haskell
awsToFinal
    :: Member (Final IO) r
    => Sem (Aws ': r) a
    -> Sem r a
awsToFinal = interpretFinal $ \case
  AwsMkEnv region cb -> do
    cb' <- bindS (uncurry cb)
    s   <- getInitialStateS
    liftS $ mkEnv region $ \loglvl str -> void $ cb' ((loglvl, str) <$ s)
```

**Comment by newhoggy** [comment]
Not quite want I wanted, but it helps a lot and I will definitely use it thanks!

I was wondering if I could avoid using an interpreter altogether when dealing with callback functions.

Partly it's because I intend to construct an `Aws` interpreter with an env, but at the moment, I can't have an env until I construct and `Aws` interpreter, so I have a chicken/egg situation

The answer might be that I have to have two interpreters in this case, one to support constructing an env and one for using it.

**Comment by newhoggy** [comment]
The context is I'm converting one of my projects to use `polysemy` to evaluate it and it uses callbacks in the setup phase, and I'm trying to figure out what's involved and also the best way to structure the app:

https://github.com/haskell-works/cabal-cache/pull/80

**Comment by alaendle** [comment]
Maybe I'm just to unfamiliar with the syntax, but how to adapt this pattern to map a nested callback?
How to use something like:
```haskell
nestedImpl :: Int -> (String -> (Bool -> IO ()) -> IO ()) -> IO ()
nestedImpl msg cb = cb (show msg) (\x -> putStrLn $ "result" ++ show x)
```
In an effect like:
```haskell
data Echo m a where
  Nested :: Int -> (String -> (Bool -> m ()) -> m ()) -> Echo m ()
```
Is my expectation/approach simply wrong? My guess is that once I got he idea right, it might be easy; but I spend a few hours finding a meaningful implementation for an interpreter - and I failed. The thing I stumble across is how to interpret the callback `Bool -> IO ()` as something of `Bool -> m ()` - so I hope I'm pretty close, but miss or misunderstood some important detail.

Thanks in advance for any hint that might lead to a solution 😉 
@newhoggy @KingoftheHomeless
